File: src/sequence/cycles.ts
import type { Factor, SeqFactor } from "../types/Sequence";
import type { IntervalType } from "./time";
import { generateTimeSeries } from "./time";

export const createCycles = (
  start: Date,
  end: Date,
  step: number,
  intervalType: IntervalType,
  factor: Factor,
  value: number,
  tag: string,
  genId: () => string,
): SeqFactor[] => {
  const times = generateTimeSeries(start, end, step, intervalType);
  const sequence: SeqFactor[] = [];
  for (const time of times) {
    sequence.push({
      id: genId(),
      tag: tag,
      value: value,
      factor: factor,
      time: time,
    });
  }
  return sequence;
};

File: cycles.go
package aggrep

import "time"

// CreateCycles creates repeated factors from start to end using a fixed interval.
func CreateCycles(
	start, end time.Time,
	step int,
	intervalType IntervalType,
	factor Factor,
	value float64,
	tag string,
	genID func() string,
) []SeqFactor {
	times := GenerateTimeSeries(start, end, step, intervalType)
	sequence := make([]SeqFactor, 0, len(times))

	for _, t := range times {
		sequence = append(sequence, SeqFactor{
			ID:     genID(),
			Tag:    tag,
			Value:  value,
			Factor: factor,
			Time:   t,
		})
	}

	return sequence
}
